internal/action/actions.go

```diff
@@ -490,38 +490,84 @@ func (h *BufPane) SelectToEndOfLine() bool {
 	return true
 }
 
-// ParagraphPrevious moves the cursor to the previous empty line, or beginning of the buffer if there's none
-func (h *BufPane) ParagraphPrevious() bool {
+func (h *BufPane) paragraphPrevious() {
 	var line int
+	// Skip to the first non-empty line
 	for line = h.Cursor.Y; line > 0; line-- {
+		if len(h.Buf.LineBytes(line)) != 0 {
+			break
+		}
+	}
+	// Find the first empty line
+	for ; line > 0; line-- {
 		if len(h.Buf.LineBytes(line)) == 0 && line != h.Cursor.Y {
 			h.Cursor.X = 0
 			h.Cursor.Y = line
 			break
 		}
 	}
-	// If no empty line found. move cursor to end of buffer
+	// If no empty line was found, move the cursor to the start of the buffer
 	if line == 0 {
 		h.Cursor.Loc = h.Buf.Start()
 	}
-	h.Relocate()
-	return true
 }
 
-// ParagraphNext moves the cursor to the next empty line, or end of the buffer if there's none
-func (h *BufPane) ParagraphNext() bool {
+func (h *BufPane) paragraphNext() {
 	var line int
+	// Skip to the first non-empty line
 	for line = h.Cursor.Y; line < h.Buf.LinesNum(); line++ {
+		if len(h.Buf.LineBytes(line)) != 0 {
+			break
+		}
+	}
+	// Find the first empty line
+	for ; line < h.Buf.LinesNum(); line++ {
 		if len(h.Buf.LineBytes(line)) == 0 && line != h.Cursor.Y {
 			h.Cursor.X = 0
 			h.Cursor.Y = line
 			break
 		}
 	}
-	// If no empty line found. move cursor to end of buffer
+	// If no empty line was found, move the cursor to the end of the buffer
 	if line == h.Buf.LinesNum() {
 		h.Cursor.Loc = h.Buf.End()
 	}
+}
+
+// ParagraphPrevious moves the cursor to the first empty line that comes before the paragraph closest to the cursor, or beginning of the buffer if there isn't a paragraph
```

Comment:
Just noticed: these comment lines are way too long, could you split them?
Besides that, the PR looks good to me.

Reply:
done